RealTime Analysis Pipeline and Experimental features refactor/additions

FilipCon opened the pull request:
This PR contains mainly the RealTime analysis pipeline class, reworked from v3.3. I removed the thread-safe implementation of the LowPassSmoothFilter, since I didn't use it in the pipeline in the end. Instead, I kept the CircularBuffer and performed the filtering in the acquisition thread.

In addition, I reorganized the classes of the experimental features (GRFM prediction, marker reconstruction) in separate folders, I added GaitPhaseDetectors that use external measurements (from sensors) to determine gait phase, and an extended version of the RealTimeAnalysis class that includes the aforementioned features.

I didn't manage to separate the Applications successfully. It can be performed in separate PR if you agree. However, the present PR concludes all my major work, so it can go public after the merge.

FilipCon committed:
merge master

mitkof6 commented on OpenSimRT/RealTime/experimental/RealTimeAnalysisExtended/include/RealTimeAnalysisExtended.h:
```diff
@@ -0,0 +1,109 @@
+ * reconstruction IK and filtering, and one processing thread for the rest the
+ * analysis (GRF&M prediction, ID, SO and JR).
+ */
+class RealTime_API RealTimeAnalysisExtended : public RealTimeAnalysis {
```
Is it necessary to extend from RealTimeAnalysis? Maybe we can decouple them.

FilipCon replied:
I did it for code reuse. Practically only the added modules and their parameters differ.

mitkof6 commented on OpenSimRT/RealTime/experimental/RealTimeAnalysisExtended/src/RealTimeAnalysisExtended.cpp:
```diff
@@ -0,0 +1,192 @@
+                externalWrenches = {grfRightWrench, grfLeftWrench};
+            } else {
+                // use filtered external measured wrenches
+                externalWrenches = filteredData.externalWrenches;
```
In the constructor: ensure that useGRFMPrediction = false when external wrenches are provided.

FilipCon replied:
I already check if the `parameters.useGRFMPrediction` is true/false. I don't see how I can check in the constructor if external wrenches are provided. The `wrenchParameters` can be used for JR as well, so there is no point of check it.

mitkof6 commented on OpenSimRT/RealTime/experimental/RealTimeAnalysisExtended/tests/TestRTExtFromFile.cpp:
```diff
@@ -0,0 +1,320 @@
+ *
+ * @file TestRTExtFromFile.cpp
+ *
+ * @brief Tests the RealTimeAnalysisExtended class with data acquired from file.
```
It is good to comment on the delay that you observed in your tests. We might need to run this on a faster computer because of the GRF prediction that might delay the pipeline.

FilipCon replied:
The GRF prediction is quite fast. I've measured the performance of the pipelines, and there is no significant difference in delay(FPS) between to original and extended pipeline. However, the comparison is not very reliable as thread priorities and system load affects heavily the measurements. Both pipelines result in ~30ms delay (without SO+JR).

mitkof6 replied:
I think 30ms is a lot without SO. What is causing this delay? How come GRF prediction/not does not affect the pipeline? Do the unit tests indicate a delay of 10ms or is it due to the streamer?

FilipCon replied:
The total processing in the acquisition thread lasts ~13ms. This is added on-top of the delay in the `dataAcquisitionFunction` passed to the pipeline which is 16ms. Resulting in 29ms delay even without the SO.

FilipCon replied:
I used to have a separate `streaming` thread for the streaming data in a constant rate, and the `dataAcquisitionFunction` would only fetch data from that stream, without the added delay. Back then, however, it was problematic to open an additional thread for streaming. I'll look back into it now that I have changed OS.

mitkof6 replied:
I remember that the pipeline was very efficient in v3.3, even with SO. We have to look at how to redesign it so that everything is efficient.

FilipCon replied:
![image](https://user-images.githubusercontent.com/26690578/111364697-3e1b6380-869a-11eb-835e-bd0069c4bf53.png)

FilipCon replied:
All the processing goes to filtering (due to splines) and SO. I don't know if those can be improved any further.

FilipCon replied:
![image](https://user-images.githubusercontent.com/26690578/111365640-6788bf00-869b-11eb-9c31-92d7133bca3a.png)

And here's another without the SO.

mitkof6 replied:
![1615946083769](https://user-images.githubusercontent.com/4024976/111402878-47bdbf00-86cc-11eb-9236-c3d75455dc05.jpg)

I have put worst-case numbers. In this case, the acquisition should have an upper limit of 140 FPS and the processing thread 45 FPS. Ensure that data comes to the acquisition thread at frames lower than 140 (e.g., 100 is a good choice). You only have to process the last frame in the processing thread since we do not have any state.

If we move the filter to the processing thread, then the acquisition thread can achieve high FPS (e.g., >500), but the processing will be around 30 FPS. It is best if we can keep the filter on the acquisition thread.

mitkof6 replied:
Then, we also have to minimize the read/write time to the buffer so that filtered data can be easily mapped to and from.

mitkof6 replied:
> All the processing goes to filtering (due to splines) and SO. I don't know if those can be improved any further.

This is a good point. Maybe in the future, we can improve the spline and filtering to reuse data from the previous step. For now, let us assume that the filter operates at 6ms worst-case and we may try to improve this.

FilipCon replied:
That's exactly the pipeline followed right now. But you forget the `dataAcquisitionFunction` is a function that lasts 16ms, which are added to the acquistion time, resulting in max 43fps. In my system I don't have the same measured delays as shown in the picture, so I get around 30ms instead.

mitkof6 replied:
OK, I see, but you should not account for this delay. The acquisition thread should execute quickly if you do not add this delay. What is then the execution time of the processing thread with SO? This will dictate the throughput of the system.

FilipCon replied:
Around 43ms.

mitkof6 replied:
Why 43? Can you specify the delay of each operation? I would expect SO to be around 20. The rest should execute fast.

mitkof6 commented on OpenSimRT/RealTime/experimental/RealTimeAnalysisExtended/tests/TestRTExtFromFile.cpp:
```diff
@@ -0,0 +1,320 @@
+    auto memoryHistory = ini.getInteger(section, "MEMORY_HISTORY", 0);
+    auto maximumIterations = ini.getInteger(section, "MAXIMUM_ITERATIONS", 0);
+    auto objectiveExponent = ini.getInteger(section, "OBJECTIVE_EXPONENT", 0);
+    auto momentArmLibraryPath =
```
It is strange that this passes the ctest without the absolute path.

FilipCon replied:
No clue...

mitkof6 commented on OpenSimRT/RealTime/experimental/RealTimeAnalysisExtended/tests/TestRTExtFromFile.cpp:
```diff
@@ -0,0 +1,320 @@
+        double t = input.IkFrame.t;
+
+        // dummy delay to simulate real time
+        this_thread::sleep_for(chrono::milliseconds(16));
```
Why 16ms and not 10?

FilipCon replied:
The dataset is not resampled at 100Hz. The datarate in the .trc file is 60Hz. Although it is not consistent...

mitkof6 commented on OpenSimRT/RealTime/include/RealTimeAnalysis.h:
```diff
@@ -0,0 +1,234 @@
+ * performing musculoskeletal analysis in mocap scenarios.
+ *
+ * @author Dimitar Stanev <jimstanev@gmail.com>
+ * contribution: Filip Konstantinos <filip.k@ece.upatras.gr>
```
Maybe remove the contribution keyward and add your name next to mine.

mitkof6 commented on OpenSimRT/RealTime/include/RealTimeAnalysis.h:
```diff
@@ -0,0 +1,234 @@
+#include "JointReaction.h"
+#include "RealTimeAnalysis.h"
+#include "internal/RealTimeExports.h"
+
```
I do not like that clang-formatter leaves space in the include. Do you think we should configure it to keep all includes together?

FilipCon replied:
I'll check. But personally I don't mind it. It (usually) separates the `"project-headers.h"` from the `<library-headers>`

mitkof6 replied:
It is good that it sorts things, but let see if it avoids adding new lines.

FilipCon replied:
It's an option of the clangd formatter. `IncludeBlocks:Merge`. https://clang.llvm.org/docs/ClangFormatStyleOptions.html

mitkof6 commented on OpenSimRT/RealTime/include/RealTimeAnalysis.h:
```diff
@@ -0,0 +1,234 @@
+ */
+class RealTime_API RealTimeAnalysis {
+ public:
+    struct UnfilteredData {
```
This was implemented initially due to the time constraints of the review. I was thinking that it would be best if we could view this as a continuous array of values or separately to avoid converting the vectors. Maybe a union or a view can work. Do you have any other ideas?

mitkof6 replied:
E.g., to be able to access t, q, w or get the whole thing as a single array without any additional computational cost.

FilipCon replied:
Since the filter requires a `SimTK::Vector` as input, the only thing I can think of is to remove the `UnfilteredData` struct entirely and keep just a function identical to `RealTimeAnalysis::UnfilteredData::toVector()` to create the input vector. This will remove the intermediate step of instantiating `UnfilteredData` objects in every loop, copying data to the struct fields and then applying `RealTimeAnalysis::UnfilteredData::toVector()`.

mitkof6 commented on OpenSimRT/RealTime/include/RealTimeAnalysis.h:
```diff
@@ -0,0 +1,234 @@
+        MomentArmFunctionT momentArmFunction;
+    };
+
+    struct Loggers {
```
Do we need the loggers here?

FilipCon replied:
I have restricted the access to the individual modules. So logger initialization can't happen from the user code. I thought it might be more compact if the user code only calls `pipeline.initializeLoggers()` once.

mitkof6 commented on OpenSimRT/RealTime/src/RealTimeAnalysis.cpp:
```diff
@@ -0,0 +1,247 @@
+using namespace OpenSimRT;
+using namespace SimTK;
+
+Vector RealTimeAnalysis::UnfilteredData::toVector() {
```
This is what I want to avoid if possible.

mitkof6 commented on OpenSimRT/RealTime/src/RealTimeAnalysis.cpp:
```diff
@@ -0,0 +1,247 @@
+    auto wrenchSize = ExternalWrench::Input::size();
+    int wrenchCount = (x.size() - nq) / wrenchSize;
+
+    ExternalWrench::Input wrench;
```
This as well

mitkof6 commented on OpenSimRT/RealTime/tests/TestRTFromFile.cpp:
```diff
@@ -0,0 +1,255 @@
+ * @brief Tests the RealTimeAnalysis class with data acquired from file.
+ *
+ * @author Dimitar Stanev <jimstanev@gmail.com>
+ * contribution: Filip Konstantinos <filip.k@ece.upatras.gr>
```
same

mitkof6 commented on OpenSimRT/RealTime/tests/TestRTFromFile.cpp:
```diff
@@ -0,0 +1,255 @@
+ *
+ * @file TestRTFromFile.cpp
+ *
+ * @brief Tests the RealTimeAnalysis class with data acquired from file.
```
Can you comment on the performance here?

mitkof6 commented on OpenSimRT/RealTime/tests/TestRTFromFile.cpp:
```diff
@@ -0,0 +1,255 @@
+        input.ExternalWrenches = {grfRightWrench, grfLeftWrench};
+
+        // dummy delay to simulate real time
+        this_thread::sleep_for(chrono::milliseconds(16));
```
Same

mitkof6 replied:
100 FPS should be fine.

mitkof6 commented on OpenSimRT/RealTime/tests/TestRTFromFile.cpp:
```diff
@@ -0,0 +1,255 @@
+
+    // initialize filter parameters
+    LowPassSmoothFilter::Parameters filterParameters;
+    filterParameters.numSignals = dofs + 18;
```
comment on 18

mitkof6 commented on OpenSimRT/RealTime/tests/TestRTFromFile.cpp:
```diff
@@ -0,0 +1,255 @@
+    pipelineParameters.solveMuscleOptimization = solveMuscleOptimization;
+    pipelineParameters.ikMarkerTasks = markerTasks;
+    pipelineParameters.ikConstraintsWeight = SimTK::Infinity;
+    pipelineParameters.ikAccuracy = 1e-5;
```
Maybe this is could be part of the ini

mitkof6 commented on OpenSimRT/RealTime/tests/TestRTFromFile.cpp:
```diff
@@ -0,0 +1,255 @@
+        pipeline.shouldTerminate(true);
+    }
+
+    // // store results
```
Maybe we have a folder in real_time and we do the comparison to ensure everything works properly.

FilipCon replied:
Comparison of the results is not feasible. Due to the use of threads the method is not deterministic and the results are not consistent between runs. I tried it and it failed. This was my main concern regarding if these files should be tests or solely applications.

mitkof6 replied:
Maybe a qualitative check in case the FPS is lower.

mitkof6 replied:
Just to validate that we did not mess things. You could adopt a python script to plot and compare results.

FilipCon replied:
That would be a lot of plots...

FilipCon replied:
I tried to rework the `compareTables` function to compare rows (as an option) instead of columns only, in order to compare common row entries in the tables. However I noticed that the results of the modules in the processing thread depend on the input value history, and since the results are not passed to the processing thread in a deterministic manner, the RMSE often exceeds the given threshold.

mitkof6 replied:
Results are placed in the buffer. It might be possible to lower the FPS of the acquisition thread so that the processing thread does not miss any frame. However, this depends on the machine being used for testing. We could have this as a test to verify that everything works but exclude the comparison afterward.

mitkof6 reviewed (changes requested):
Seems nice. Thanks! Can you check:

1. Why dt is not constant in the filter? This is important.
2. Refactor the external.
3. Check the performance.

FilipCon commented on OpenSimRT/Common/include/CircularBuffer.h:
```diff
@@ -45,6 +46,7 @@ template <int history, typename T> class CircularBuffer {
         current = 0;
         startOver = false;
         newValue = false;
+        externalNotification = false;
```
This is to notify the consumer thread and remove the `conditional_variable` from the wait state in case where no new data are added to the buffer. It was added to terminate the program normally when the simulation ends. I've added a comment mentioning this.

FilipCon commented on OpenSimRT/Common/src/SignalProcessing.cpp:
```diff
@@ -145,7 +145,7 @@ LowPassSmoothFilter::filter(const LowPassSmoothFilter::Input& input) {
 
     // check if dt is consistent
     if (abs(dt - dtPrev) > 1e-5) {
-        THROW_EXCEPTION("signal sampling frequency is not constant");
```
Yes. This was not meant to be commented. However the experimental data of the `gait1992` do not have consistent sampling rate, thus this condition becomes true and the simulation fails.

FilipCon replied:
Two ways I can think for fixing this are either using another (new or resampled) dataset only for the Pipeline tests, or resampling the existing dataset through the SyncManager during the simulation. The second may require a littler more effort.

mitkof6 replied:
When I load the data set I resample it. But, I think the marker frame rate should be constant. Maybe the error comes from somewhere else.

FilipCon replied:
Then the error should be originated from the acquisition thread not being able to "pick" every sample in time. The extra processing time due to filtering in the acquisition thread probably results in lost frames.

mitkof6 replied:
But the FPS of the acquisition thread should be high (>500 frames).

FilipCon replied:
Turns out that is just a rounding error due to the datarate of the data. The 60Hz result in `dt` that is either 16ms or 17ms. And that's why the check `abs(dt - dtPrev) > 1e-5` fails.

mitkof6 replied:
Maybe comment on this in the code so that we remember this point.

FilipCon commented on OpenSimRT/Common/src/Visualization.cpp:
```diff
@@ -114,6 +114,7 @@ void BasicModelVisualizer::update(const Vector& q,
 #endif
     // kinematics
     state.updQ() = q;
+    model.realizePosition(state);
```
I got errors on the State not being updated properly when visualizing the joint reaction loads. This seemed to fix the issue.

```
  Error detected by Simbody method getCacheEntry: State Cache entry was out of date at Stage Time. This entry depends on version 1 of Stage Instance but was last updated at version 0.
  (Required condition 'version == m_dependsOnVersionWhenLastComputed' was not met.)
```

FilipCon commented on OpenSimRT/IMU/src/TimeConversion.cpp:
```diff
@@ -17,7 +17,7 @@
  * -----------------------------------------------------------------------------
  */
-#include "IMU/include/TimeConversion.h"
+#include "TimeConversion.h"
```
I use the tool for auto-imported headers. The previous path must have been imported automatically by the tool.

mitkof6 replied:
Which package?

FilipCon replied:
lsp+clangd

FilipCon commented on OpenSimRT/RealTime/experimental/GroundWrenchPrediction/include/AccelerationBasedPhaseDetector.h:
```diff
@@ -29,7 +29,6 @@
 #include "GRFMPrediction.h"
 #include "GaitPhaseDetector.h"
 #include "SignalProcessing.h"
-#include "internal/RealTimeExports.h"
```
It is already imported in the `GaitPhaseDetector.h`.

FilipCon committed:
review changes. refactor, dt error, performance, rm unfiltered data

mitkof6 commented on OpenSimRT/Common/include/CircularBuffer.h:
```diff
@@ -45,9 +46,15 @@ template <int history, typename T> class CircularBuffer {
         current = 0;
         startOver = false;
         newValue = false;
+        externalNotification = false; // flag to remove the condition_variable
```
In the docs of this class, we should summarize how this buffer works with examples. Show different scenarios of consumer-producer based on the flags. For example, is the buffer blocking the consumer thread and how the notification/synchronization works.

mitkof6 commented on OpenSimRT/Common/include/OpenSimUtils.h:
```diff
@@ -74,27 +74,27 @@ struct Common_API OpenSimUtils {
-            THROW_EXCEPTION("Number of columns in given tables do not match.");
-
+                              const double& threshold = 1e-5,
+                              const bool& compareColumns = true) {
```
Should verify that previous test work properly.

FilipCon replied:
Tests work and pass as expected.

mitkof6 commented on OpenSimRT/RealTime/CMakeLists.txt:
```diff
@@ -2,23 +2,41 @@
 include_directories(../Common/include/)
 set(DEPENDENCY_LIBRARIES ${OpenSim_LIBRARIES} Common)
 
+if(USE_EXPERIMENTAL)
```
Remove USE_EXPERIMENTAL and build by default.

mitkof6 commented on OpenSimRT/RealTime/CMakeLists.txt:
```diff
@@ -2,23 +2,41 @@
 
+if(USE_EXPERIMENTAL)
+  # global experimental files
+  file(GLOB add_includes include/experimental/*.h)
```
This could be moved as part of the sources and includes to make things simple.

mitkof6 commented on OpenSimRT/RealTime/CMakeLists.txt:
```diff
@@ -2,23 +2,41 @@
+  file(GLOB add_sources src/experimental/*.cpp)
+  # file(GLOB add_tests tests/experimental/*.cpp)
+  file(GLOB add_tests
+    tests/experimental/TestAccelerationGRFMPredictionFromFile.cpp
```
You can move these to tests.

mitkof6 commented on OpenSimRT/RealTime/CMakeLists.txt:
```diff
@@ -2,23 +2,41 @@
+    )
+
+  # append files
+  set(includes ${includes} ${add_includes})
```
Same, no need to have includes and add_includes.

mitkof6 commented on OpenSimRT/RealTime/CMakeLists.txt:
```diff
@@ -35,9 +53,3 @@ addTests(
   LINKLIBS ${target} ${DEPENDENCY_LIBRARIES}
   )
-
-# applications
```
We do not have applications in this module, right?

FilipCon replied:
No, we don't.

mitkof6 commented on OpenSimRT/RealTime/include/InverseDynamics.h:
```diff
@@ -51,7 +51,7 @@ class RealTime_API ExternalWrench : public OpenSim::Force {
         SimTK::Vec3 force;
         SimTK::Vec3 torque;
-        SimTK::Vector toVector();
+        SimTK::Vector toVector() const;
```
It is good to check if we need to add const to other functions as well.

mitkof6 commented on OpenSimRT/Common/include/CircularBuffer.h:
```diff
@@ -22,13 +22,14 @@
  *
  * @author Dimitar Stanev <jimstanev@gmail.com>
- * @contribution Filip Konstantinos <filip.k@ece.upatras.gr>
+ * contribution: Filip Konstantinos <filip.k@ece.upatras.gr>
```
Put your name next to mine.

mitkof6 commented on OpenSimRT/RealTime/include/RealTimeAnalysis.h:
```diff
@@ -0,0 +1,224 @@
+    SimTK::ReferencePtr<JointReaction> jointReaction;
+
+    // data buffer
+    CircularBuffer<1, LowPassSmoothFilter::Output> buffer;
```
Maybe we have a larger buffer in case we change the behavior in the future, but process only the last frame in the processing thread. What do you think?

FilipCon replied:
Holds resources for no reason. I don't see the point in this. I could leave a note, or create a #define with the buffer size.

mitkof6 commented on OpenSimRT/RealTime/src/RealTimeAnalysis.cpp:
```diff
@@ -0,0 +1,243 @@
+
+            // get data
+            auto acquisitionData = parameters.dataAcquisitionFunction();
+            if (previousAcquisitionTime >= acquisitionData.IkFrame.t) {
```
Maybe we should place things in a data buffer and notify this thread when we have a new frame.

mitkof6 replied:
In this case, it is important that the acquisition thread has a higher FPS than the acquisition system to ensure that we consume everything with a constant dt for filtering.

mitkof6 commented on OpenSimRT/RealTime/src/RealTimeAnalysis.cpp:
```diff
@@ -0,0 +1,243 @@
+
+            // get data from buffer
+            auto data = buffer.get(1)[0];
+            filteredData.fromVector(data.t, data.x, data.xDot, data.xDDot,
```
To speed up things in data transfer, maybe we can use pointers to point to the new vectors.

mitkof6 commented on OpenSimRT/RealTime/src/experimental/ExternalAccelerationBasedPhaseDetector.cpp:
```diff
@@ -0,0 +1,60 @@
+ * -----------------------------------------------------------------------------
+ */
+
+#include "ExternalAccelerationBasedPhaseDetector.h"
```
Why Extrernal? Maybe use a shorter name.

FilipCon replied:
The 'external' was meant as an indication that the acceleration values that dictate the gait phase are measured from an external sensor, instead of being estimations from the model after IK (internal acceleration). Perhaps 'Measured' would be a better name?

mitkof6 commented on OpenSimRT/RealTime/src/experimental/RealTimeAnalysisExtended.cpp:
```diff
@@ -0,0 +1,189 @@
+ * OpenSimRT. If not, see <https://www.gnu.org/licenses/>.
+ * -----------------------------------------------------------------------------
+ */
+#include "RealTimeAnalysisExtended.h"
```
Maybe think of a better name for this pipeline. Also, it might be better to give another name to RealTimeAnalysis as well. Can you propose something?

FilipCon replied:
Not really.. It's based on the Facade design pattern, so I'd only suggest a name that includes this, so it is more obvious to the user. Like, 'PipelineAnalysisFacade' or similar..

FilipCon committed:
reformat all files

FilipCon committed:
refactor experimental features, test performance

FilipCon committed:
update docs

mitkof6 commented on OpenSimRT/Common/include/CircularBuffer.h:
```diff
@@ -38,7 +37,29 @@
+ *  inserted in the buffer. Retrieval is possible only if new elements have been
+ *  appended to the buffer. To retrieve data from the buffer (and terminate the
+ *  'wait' state of the consumer thread) even if no new data have been added,
+ *  the user must first call the `externalNotify()` function, which notifies the
```
Try to give an example of the thread-safety and the externalNotify through code after the diagram, so it is clearer. It is not very clear what you mean "to retrieve data from the buffer even if no new data have been added.

Also, do you think externalNotify is a good name? How does it work?

mitkof6 commented on OpenSimRT/Common/src/SignalProcessing.cpp:
```diff
@@ -144,7 +142,7 @@ LowPassSmoothFilter::filter(const LowPassSmoothFilter::Input& input) {
 
     // check if dt is consistent
-    if (abs(dt - dtPrev) > 1.5e-3) {
+    if (abs(dt - dtPrev) > 1.5e-3) { // NOTE: ensure double precision if throws
```
Try to explain that this value works well in cases of 60Hz (or 0.016) which can sometimes due to floating-point errors become 0.017. It is still strange that we have such a large error on the 3rd digit.

mitkof6 commented on OpenSimRT/RealTime/CMakeLists.txt:
```diff
@@ -9,34 +11,18 @@ file(GLOB tests
   tests/TestSOFromFile.cpp
   tests/TestJRFromFile.cpp
   tests/TestRTFromFile.cpp
+  tests/TestAccelerationGRFMPredictionFromFile.cpp
```
Maybe put tests these into an experimental folder.

mitkof6 commented on OpenSimRT/RealTime/include/experimental/ExternalAccelerationBasedPhaseDetector.h:
```diff
@@ -33,7 +33,6 @@
 #include "GaitPhaseDetector.h"
```
Sorry, I do not get it. Why do we have AccelerationBasedPhaseDetector and ExternalAccelerationBasedPhase detector?

mitkof6 replied:
I will have to check the code, but do you think we can merge the two?

FilipCon replied:
The first estimates the foot acceleration from station points in the model, and the second uses actual  acceleration measurements from sensors. It could be merged, but I have to take a second look, because it might get more complicated that it needs to be.

mitkof6 commented on OpenSimRT/RealTime/tests/TestRTExtFromFile.cpp:
```diff
@@ -20,7 +20,7 @@
  *
  * @brief Tests the RealTimeAnalysisExtended class with data acquired from file.
- * Observed delay (with GRFMPredition) = ~30ms without SO + JR, ~42ms with
+ * Observed delay (with GRFMPredition) = ~19ms without SO + JR, ~31ms with
```
I think the variable FPS that you observed is due to thread scheduling that can mess up with how fast the acquisition thread pushes things in the buffer. To improve this, maybe we can implement a mechanism that places things at a constant FPS since the acquisition thread has this luxury.

FilipCon replied:
Or it could just be that the SO alone ranges from 8ms to 125ms for a single frame. Pushing to the data buffer is 0ms, so this cannot be an issue. Retrieving data from the buffer has an upper cap since it cannot last any longer than the acquisition rate + IK + Filter time. 

My guess would be that the samples that the SO receives are not contiguous, and that makes it harder to find a solution close to the previous estimations. But since I did not have any issues in Windows, I could be wrong. Could it be a Ipop implementation issue? Have you tried it and experienced the same behavior?

mitkof6 commented on OpenSimRT/Vicon/include/ViconDataStream.h:
```diff
@@ -32,7 +32,6 @@
 #include "CircularBuffer.h"
```
Tsaopoulos mentioned that we could create a docker that can set up this environment so that it could be run on Windows. This might be nice because we can send him the docker file and test it without being there to build the whole system. We might have to include the lower limb model that was used for the experiments. Did you change it?

FilipCon replied:
We used the same model for the experiments.

mitkof6 reviewed (commented):
Thanks, I will have to test things and come back with comments.

FilipCon committed:
experimental tests, CircularBuffer docs and test

FilipCon committed:
complete docs in CircularBuffer

mitkof6 commented on OpenSimRT/Common/CMakeLists.txt:
```diff
@@ -4,7 +4,7 @@ file(GLOB sources src/*.cpp)
 # file(GLOB tests tests/*.cpp)
 file(GLOB tests
   # Does not terminate because CircularBuffer waits when empty (new implementaiton).
```
Just a comment for me to check this test and remove the comment if it works.

mitkof6 reviewed (approved):
Thanks, I will test the branch today or tomorrow.

FilipCon committed:
remove commment. TestCircularBuffer works.

Commit:
refactor tests

Commit:
include ifndef so that tests do not fail on Windows

mitkof6 commented:
@FilipCon I included some changes to the tests. Could you please look at the following issues that I identified:

1. Frame rate on tests: IMU Upper/Lower limbs from file does not go above 30 FPS on Windows. On Linux, it seems that it works well.
2. The IMU tests upper/lower from file go into strange configuration at the end. Could you check if this could be fixed maybe by skipping the last frame?
3.  If you close the IMU from file tests for the upper and lower limb the application does not terminate. Maybe the data push thread is detached.
4. **Note**: GRF and SO tests fail on Windows. I have included ifndef clause. However, it will be nice if we can identify why. Maybe we make this an issue.
5. **IMPORTANT**: I observed that the RT pipeline messes up the muscle activations. It seems that muscles are on/off all the time. Something is wrong with the pipeline. We need to fix this before this is merged.

Commit:
probably tests were not performed

FilipCon committed:
Add MarkerTasks in pipeline. Fix NGIMUInputFromFileDriver.

mitkof6 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -83,4 +83,6 @@ jobs:
       run: |
         source $GITHUB_WORKSPACE/.github/workflows/env_variables
-        ctest --parallel -j$(nproc) -VV --output-on-failure -E TestMomentArm TestViconSDK
+        pwd
```
Could you remove the pwd, I forgot to remove it?

mitkof6 reviewed (commented):
Could you please revert back to our results on the real-time pipeline? I think the ones you generated are a bit noisy, although they do not fail the test in Ubuntu. This could be due to the max_iteration option of the muscle optimization.

FilipCon committed:
Revert to original results

FilipCon committed:
remove pwd

FilipCon commented:
> @FilipCon I included some changes to the tests. Could you please look at the following issues:
> 
> 1. Frame rate on tests: IMU Upper/Lower limbs from file does not go above 30 FPS on Windows. On Linux, it seems that it works well.
> 2. The IMU tests upper/lower from file go into strange configuration at the end. Could you check if this could be fixed maybe by skipping the last frame?
> 3. If you close the IMU from file tests for the upper and lower limb the application does not terminate. Maybe the data push thread is detached.
> 4. **Note**: GRF and SO tests fail on Windows. I have included ifndef clause. However, it will be nice if we can identify why. Maybe we make this an issue.
> 5. **IMPORTANT**: I observed that the RT pipeline messes up the muscle activations. It seems that muscles are on/off all the time. Something is wrong with the pipeline. We need to fix this before this is merged.

1. I think this is due to the simulated delay in stream and the minimal time resolution in Windows systems. 
2. Fixed it.
3. I've added the ESC key to close properly the visualizer.
4. There are errors in precision between systems, but plotted graphs show that the results are close.
5. The pipeline works fine. The issue must be in the used dataset and the double precision in time intervals when using 60Hz.

mitkof6 commented:
OK, I had a check. Most of the things seem fine. Let us try to figure out why the pipeline is so noisy.

First of all, I think the CircularBuffer has a bug. You have introduced the newValue condition, but this does not work properly when the buffer's memory is > 1.

About the RT pipeline, I am also experimenting. Could you try to test the pipeline by moving the filtering to the processing thread. Also, we can experiment with gait1848.

mitkof6 commented:
Could you then resample the file in python and export the values in a separate file? Also, for the analysis that fails, we could add a try-catch statement with a message so that the CI test does not fail due to floating-point errors.

mitkof6 commented:
>I'll have a look again one the CircularBuffer then. Have you checked the operation mode (CONTINUOUS ON_ENTRY)?

This is a bit confusing. It might be best to have a method that unblocks the other thread (if I understood the use-case).

FilipCon committed:
Resample markers and address noise issue in RT pipeline

FilipCon committed:
Revert sampling threshold in LowPassSmoothFilter

FilipCon committed:
Fix wait condition in CircularBuffer. Buffer must have M values.

mitkof6 commented on OpenSimRT/Common/include/CircularBuffer.h:
```diff
@@ -21,34 +21,99 @@
+ * - ON_ENTRY allows the user user to get data from the buffer only after
+ * new data have been added.
+ *
+ * - CONTINUOUS mode allows the user the retrieve data from the buffer
```
The continuous is a bit misleading. I would say that you always get the last data. As it is explained, I understand that it gets the value and removes it from the buffer until the buffer is empty. This is not a stack. The name is fine but maybe explain it better.

mitkof6 commented on OpenSimRT/Common/include/CircularBuffer.h:
```diff
@@ -75,15 +157,21 @@ template <int history, typename T> class CircularBuffer {
+    /**
+     * Retrieve data from buffer. To retrieve data, the buffer must have at
+     * least M elements, else the consumer thread is blocked until buffer is
+     * filled with M values.
```
Point out that the blocking behavior is changed when the model = continuous, thus the get call will always return the latest M elements regardless of whether we have a new entry.

mitkof6 commented on OpenSimRT/Common/src/SignalProcessing.cpp:
```diff
@@ -182,115 +180,6 @@ LowPassSmoothFilter::filter(const LowPassSmoothFilter::Input& input) {
-    newDataReady = false; // do not filter unless new data are entered in matrix
-
-    // set dt
-    dt = time[0][M - 1] - time[0][M - 2];
```
I know this was removed. I think the calculation of dt might have caused the problem. With this implementation, dt can change due to precision errors. Maybe, we can specify dt externally as a parameter. Can you verify this with the old and new MarkerData?

FilipCon replied:
Setting only the `dt` in the LowPassSmoothFilter won't do the trick. When pushing to the time matrix using the `shiftColumnsLeft(Vector(1, input.t), time);` you have to push time values that have constant rate. I have tested it and verified that, using the markerData @60Hz, the pipeline works as long as you replace the time values from file with static values that increment in the acquisition function passed in the pipeline parameters. Eg.:
```
    // acquisition function (simulates acquisition from motion)
    auto dataAcquisitionFunction = [&]() -> MotionCaptureInput {
        static int i = 0;
        static double t = 0.0; // <<<<<<<<<<<<<<<<<<< HERE
        
        MotionCaptureInput input;
        // get frame data
        input.IkFrame = InverseKinematics::getFrameFromMarkerData(
                i, markerData, observationOrder, false);
        input.IkFrame.t = t;

        // get grf force
        auto grfRightWrench = ExternalWrench::getWrenchFromStorage(
                t, grfRightLabels, grfMotion);
        auto grfLeftWrench = ExternalWrench::getWrenchFromStorage(
                t, grfLeftLabels, grfMotion);
        input.ExternalWrenches = {grfRightWrench, grfLeftWrench};

        // dummy delay to simulate real time
        this_thread::sleep_for(10ms);
        i++;
        t += 1 / 60.0; // <<<<<<<<<<<<<<<<<<< HERE
        return input;
    };
```

I think resampling the original data is more appropriate than error-proofing the filter in this case, since data are retrieved from file. In real-time scenarios, one can use the SyncManager (or any other future implementation) to resample the streamed data online.

mitkof6 commented on OpenSimRT/RealTime/tests/TestSOFromFile.cpp:
```diff
@@ -145,7 +154,8 @@ void run() {
-    // machine, possibly due to compilation differences.
+    // machine, possibly due to compilation differences (this fails on Windows).
+#ifndef WIN32
     try {
```
Can you do the try-catch for the RT tests so that CI does not fail?

FilipCon committed:
Improve CircularBuffer docs. Add try/catch in pipeline tests

mitkof6 reviewed (approved):
Thanks, everything seems fine. Just some minor comments and I can merge.

1. Could you add a comment for why we add the try-catch statement on the 3 test files?
2. Maybe you did it but I am not sure. Could you re-generate the test files for the RT pipelines now that you fixed the issue? Ideally, you should have high FPS and do this on Ubuntu 18.04 to avoid any kind of mismatch.

FilipCon committed:
Regenerate result files for RT pipelines. Add comment on try/catch.

mitkof6 commented:
Thanks! Results seem worst than before. I can also try to generate them on Windows where I can achieve around 30 FPS.

FilipCon commented:
On Ubuntu I get around 30-40 fps. Please note that the comparisons are performed with the data resulted after RRA!

mitkof6 commented:
Yes, apart from the pelvis, the curves are way off. In the previous version, they were much closer.

Commit:
will check regenerating results

Commit:
regenerate results for the pipeline

mitkof6 commented:
@FilipCon I regenerated the results for the pipeline and add some more fixes. Could you please check the results in the `pipeline/ext`. The graphs do not seem fine. For the normal pipeline results, I think the results are more or less fine.

mitkof6 commented:
Just pull the branch and do not rebase. Otherwise, we commit my changes twice.

FilipCon committed:
Move GRFMPrediction to acquisition thread

mitkof6 commented:
Thanks, I will merge.
